test/functional/p2p_compactblocks_blocksonly.py

```diff
@@ -0,0 +1,133 @@
+        #   1 -> high bandwidth
+        #   2 -> miner
+        #   3 -> low bandwidth
+        #
+        # Topology:
+        #   p2p_conn_blocksonly ---> node0
+        #   p2p_conn_high_bw    ---> node1
+        #   p2p_conn_low_bw     ---> node3
+        #   node2 (no connections)
+        #
+        # node2 produces blocks which get passed to the reset of the nodes
+        # through the respective p2p connections.
+
+        self.log.info("Test that blocksonly nodes do not request high bandwidth mode.")
+
+        block0 = self.build_block_on_tip()
+
+        # A blocksonly node should not request high bandwidth mode upon
+        # receiving a new valid block at the tip.
+        p2p_conn_blocksonly.send_and_ping(msg_block(block0))
+        assert_equal(int(self.nodes[0].getbestblockhash(), 16), block0.sha256)
+        p2p_conn_blocksonly.wait_until(lambda: p2p_conn_blocksonly.message_count['sendcmpct'] == 2)
+        assert_equal(p2p_conn_blocksonly.last_message['sendcmpct'].announce, False)
+
+        # A normal node participating in transaction relay should request high
+        # bandwidth mode upon receiving a new valid block at the tip.
+        p2p_conn_high_bw.send_and_ping(msg_block(block0))
+        assert_equal(int(self.nodes[1].getbestblockhash(), 16), block0.sha256)
+        p2p_conn_high_bw.wait_until(lambda: p2p_conn_high_bw.message_count['sendcmpct'] == 3)
+        assert_equal(p2p_conn_high_bw.last_message['sendcmpct'].announce, True)
+
+        # Don't send a block from the p2p_conn_low_bw so the bitcoind node
+        # doesn't select it for high bw relay.
+        self.nodes[3].submitblock(block0.serialize().hex())
+
+        self.log.info("Test that blocksonly nodes send getdata(BLOCK) instead"
+                      "of getdata(CMPCT) in low bandwidth mode.")
+
+        block1 = self.build_block_on_tip()
+
+        p2p_conn_blocksonly.send_message(msg_headers(headers=[CBlockHeader(block1)]))
+        p2p_conn_blocksonly.sync_send_with_ping()
+        assert_equal(p2p_conn_blocksonly.last_message['getdata'].inv, [CInv(MSG_BLOCK | MSG_WITNESS_FLAG, block1.sha256)])
+
+        p2p_conn_high_bw.send_message(msg_headers(headers=[CBlockHeader(block1)]))
+        p2p_conn_high_bw.sync_send_with_ping()
+        assert_equal(p2p_conn_high_bw.last_message['getdata'].inv, [CInv(MSG_CMPCT_BLOCK, block1.sha256)])
+
+        self.log.info("Test that getdata(CMPCT) is still sent on regular low bandwdith connections.")
+
+        p2p_conn_low_bw.send_and_ping(msg_headers(headers=[CBlockHeader(block1)]))
+        p2p_conn_low_bw.sync_with_ping()
+        assert_equal(p2p_conn_low_bw.last_message['getdata'].inv, [CInv(MSG_CMPCT_BLOCK, block1.sha256)])
+
+        self.log.info("Test that blocksonly nodes still serve compact blocks.")
+
+        def test_for_cmpctblock(block):
+            def test():
+                sha256 = p2p_conn_blocksonly.last_message['cmpctblock'].header_and_shortids.header.rehash()
```

Comment:
If you separate the test for existence of the `cmpctblock` message from the test of its hash, you don't need the `sync_send_with_ping()` call:

```suggestion
                if 'cmpctblock' not in p2p_conn_blocksonly.last_message:
                    return False
                sha256 = p2p_conn_blocksonly.last_message['cmpctblock'].header_and_shortids.header.rehash()
```

I'd also suggest using a lambda so that you don't need to have a named function returning the inner function, and also parametrizing the connection. Full diff:

```diff
         self.log.info("Test that blocksonly nodes still serve compact blocks.")
 
-        def test_for_cmpctblock(block):
-            def test():
-                sha256 = p2p_conn_blocksonly.last_message['cmpctblock'].header_and_shortids.header.rehash()
-                return sha256 == block.sha256
-            return test
+        def test_for_cmpctblock(conn, block):
+            if 'cmpctblock' not in conn.last_message:
+                return False
+            return conn.last_message['cmpctblock'].header_and_shortids.header.rehash() == block.sha256
 
         p2p_conn_blocksonly.send_message(msg_getdata([CInv(MSG_CMPCT_BLOCK, block0.sha256)]))
-        p2p_conn_blocksonly.sync_send_with_ping()
-        p2p_conn_blocksonly.wait_until(test_for_cmpctblock(block0))
+        p2p_conn_blocksonly.wait_until(lambda: test_for_cmpctblock(p2p_conn_blocksonly, block0))
 
         # Request high bandwidth mode from node0
         p2p_conn_blocksonly.send_and_ping(msg_sendcmpct(announce=True, version=2))
@@ -127,7 +123,7 @@ class P2PCompactBlocksBlocksOnly(BitcoinTestFramework):
         block2 = self.build_block_on_tip()
         self.nodes[0].submitblock(block1.serialize().hex())
         self.nodes[0].submitblock(block2.serialize().hex())
-        p2p_conn_blocksonly.wait_until(test_for_cmpctblock(block2))
+        p2p_conn_blocksonly.wait_until(lambda: test_for_cmpctblock(p2p_conn_blocksonly, block2))
```